xarray/core/indexing.py

```diff
@@ -1380,13 +1386,21 @@ def __getitem__(self, key):
                 return value
 
     def __setitem__(self, key, value):
-        raise TypeError(
-            "this variable's data is stored in a dask array, "
-            "which does not support item assignment. To "
-            "assign to this variable, you must first load it "
-            "into memory explicitly using the .load() "
-            "method or accessing its .values attribute."
-        )
+        if LooseVersion(dask_version) >= LooseVersion("2021.04.0+17"):
+            if isinstance(key, BasicIndexer):
+                self.array[key.tuple] = value
+            elif isinstance(key, VectorizedIndexer):
+                self.array.vindex[key.tuple] = value
+            elif isinstance(key, OuterIndexer):
+                self.array[key.tuple] = value
```

Comment:
Outer indexing has different semantics than NumPy. Namely, indexing a 2D array like `x[i, j]` where `i` and `j` are 1D arrays would override every element that matches _either_ `i` or `j` (a 2D sub-matrix) rather than the elements that match _both_ `i` and `j` (the diagonal).

Are you sure Dask implements outer indexing currently for setitem, and intends to support it in the future? This would surprise me.

Reply:
No. As I said I don't quite understand the inner workings of xarray and I don't fully understand what outer indexing is. How do you suggest an outer indexer is handled?

Reply:
No worries :).

For now I would suggest either:
1. Just raise `NotImplementedError`. We can figure out how to handle this properly later, or
2. Only handle the case where this is at most one array index that is not a `slice`, e.g.,
```
num_non_slices = sum(0 if isinstance(k, slice) else 1 for k in key.tuple)
if num_non_slices > 1:
    raise NotImplementedError("xarray can't set arrays with multiple array indices to dask yet.")
self.array[key.tuple] = value
```

Reply:
Copied verbatim.